File: backend/app/schemas/quiz.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional, Union

from pydantic import BaseModel

Difficulty = Literal["easy", "medium", "hard"]
QuestionType = Literal["mcq", "tf"]
QuizStatusLit = Literal["pending", "generating", "ready", "failed", "submitted"]


# ---------------------------------------------------------------------------
# Request shapes
# ---------------------------------------------------------------------------

class GenerateQuizRequest(BaseModel):
    num_questions: Literal[5, 10, 15] = 10
    difficulty: Optional[Difficulty] = None


class SubmitAnswer(BaseModel):
    question_id: int
    answer: Union[int, bool]


class SubmitQuizRequest(BaseModel):
    answers: list[SubmitAnswer]


# ---------------------------------------------------------------------------
# Response shapes
# ---------------------------------------------------------------------------

class GenerateQuizResponse(BaseModel):
    quiz_id: int
    status: QuizStatusLit


class QuestionPublic(BaseModel):
    id: int
    type: QuestionType
    prompt: str
    options: Optional[list[str]] = None


class AnswerReview(BaseModel):
    question_id: int
    type: QuestionType
    prompt: str
    options: Optional[list[str]] = None
    user_answer: Optional[Union[int, bool]] = None
    correct_answer: Union[int, bool]
    correct: bool
    explanation: str


class AttemptResult(BaseModel):
    attempt_id: int
    quiz_id: int
    score: float
    correct_count: int
    total_count: int
    submitted_at: datetime
    review: list[AnswerReview]


class QuizDetail(BaseModel):
    quiz_id: int
    status: QuizStatusLit
    difficulty: Difficulty
    difficulty_source: str
    num_questions: int
    created_at: datetime
    ready_at: Optional[datetime] = None
    questions: Optional[list[QuestionPublic]] = None
    attempt: Optional[AttemptResult] = None
    error_msg: Optional[str] = None


class QuizHistoryItem(BaseModel):
    quiz_id: int
    difficulty: Difficulty
    num_questions: int
    score: Optional[float] = None
    correct_count: Optional[int] = None
    submitted_at: Optional[datetime] = None
    created_at: datetime
    status: QuizStatusLit


class QuizHistoryResponse(BaseModel):
    items: list[QuizHistoryItem]
    has_open_quiz: bool
    open_quiz_id: Optional[int] = None

File: frontend/api/quiz.ts
import API from "@/api/axios"

export type Difficulty = "easy" | "medium" | "hard"
export type QuizStatus = "pending" | "generating" | "ready" | "failed" | "submitted"
export type QuestionType = "mcq" | "tf"

export interface QuestionPublic {
  id: number
  type: QuestionType
  prompt: string
  options?: string[]
}

export interface AnswerReview {
  question_id: number
  type: QuestionType
  prompt: string
  options?: string[]
  user_answer?: number | boolean | null
  correct_answer: number | boolean
  correct: boolean
  explanation: string
}

export interface AttemptResult {
  attempt_id: number
  quiz_id: number
  score: number
  correct_count: number
  total_count: number
  submitted_at: string
  review: AnswerReview[]
}

export interface QuizDetail {
  quiz_id: number
  status: QuizStatus
  difficulty: Difficulty
  difficulty_source: string
  num_questions: number
  created_at: string
  ready_at?: string | null
  questions?: QuestionPublic[] | null
  attempt?: AttemptResult | null
  error_msg?: string | null
}

export interface QuizHistoryItem {
  quiz_id: number
  difficulty: Difficulty
  num_questions: number
  score?: number | null
  correct_count?: number | null
  submitted_at?: string | null
  created_at: string
  status: QuizStatus
}

export interface QuizHistoryResponse {
  items: QuizHistoryItem[]
  has_open_quiz: boolean
  open_quiz_id?: number | null
}

export async function generateQuiz(
  fileId: number,
  body: { num_questions: 5 | 10 | 15; difficulty?: Difficulty }
): Promise<{ quiz_id: number; status: QuizStatus }> {
  const res = await API.post(`/quiz/files/${fileId}/generate`, body)
  return res.data
}

export async function getQuiz(quizId: number): Promise<QuizDetail> {
  const res = await API.get(`/quiz/${quizId}`)
  return res.data
}

export async function submitQuiz(
  quizId: number,
  answers: { question_id: number; answer: number | boolean }[]
): Promise<AttemptResult> {
  const res = await API.post(`/quiz/${quizId}/submit`, { answers })
  return res.data
}

export async function getQuizHistory(fileId: number): Promise<QuizHistoryResponse> {
  const res = await API.get(`/quiz/files/${fileId}/history`)
  return res.data
}
